src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,166 @@
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkPositionIndex(index);
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> current = nodePosition(index);
+        if (current == head) {
+            Node<T> newNode = new Node<T>(null, value, current);
+            current.prev = newNode;
+            head = newNode;
+        } else {
+            Node<T> newNode = new Node<T>(current.prev, value, current);
+            current.prev.next = newNode;
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return nodePosition(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> current = nodePosition(index);
+        T initial = current.element;
+        current.element = value;
+        return initial;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(nodePosition(index));
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (current.element == object
+                    || (current.element != null && current.element.equals(object))) {
+                unlink(nodePosition(i));
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private T unlink(Node<T> node) {
+        size--;
+        if (head == node && tail == node) {
+            head = null;
+            tail = null;
+            return node.element;
+        }
+        if (head == node) {
+            node.next.prev = null;
+            head = node.next;
+            return node.element;
+        }
+        if (tail == node) {
+            node.prev.next = null;
+            tail = node.prev;
+            return node.element;
+        }
+        node.prev.next = node.next;
+        node.next.prev = node.prev;
+        return node.element;
+    }
+
+    private void checkElementIndex(int index) {
+        if (index < 0 || index >= size) {
+            throwException(index);
+        }
+    }
+
+    private void checkPositionIndex(int index) {
+        if (index < 0 || index > size) {
+            throwException(index);
+        }
+    }
+
+    private void throwException(int index) {
+        throw new IndexOutOfBoundsException("Index: " + index
```

Comment:
uninformative method name. what if we add another method with another exception?